src/ops/indexing.rs

```diff
@@ -0,0 +1,1799 @@
+                    Slice { .. } => remaining_indices.push(ArrayIndexOp::range_full()),
+                    _ => {}
+                }
+            }
+
+            // append the remaining operations
+            remaining_indices.extend(operations[(last_array_or_index + 1)..].iter().cloned());
+        } else {
+            // !gather_first
+            for item in operations.iter() {
+                match item {
+                    TakeIndex { .. } | TakeArray { .. } => break,
+                    ExpandDims => remaining_indices.push(item.clone()),
+                    _ => remaining_indices.push(ArrayIndexOp::range_full()),
+                }
+            }
+
+            // handle the trailing gathers
+            remaining_indices.extend((0..max_dims).map(|_| ArrayIndexOp::range_full()));
+
+            // and the remaining operations
+            remaining_indices.extend(operations[(last_array_or_index + 1)..].iter().cloned());
+        }
+    }
+
+    if have_array && remaining_indices.is_empty() {
+        // `clone` returns a new array with the same shape and data
+        match src {
+            Cow::Borrowed(src) => return src.clone(),
+            Cow::Owned(src) => return src,
+        }
+    }
+
+    if remaining_indices.is_empty() {
+        remaining_indices = operations.to_vec();
+    }
+
+    // Slice handling
+    let ndim = src.ndim();
+    let mut starts: SmallVec<[i32; 4]> = smallvec![0; ndim];
+    let mut ends: SmallVec<[i32; 4]> = SmallVec::from_slice(src.shape());
+    let mut strides: SmallVec<[i32; 4]> = smallvec![1; ndim];
+    let mut squeeze_needed = false;
+    let mut axis = 0;
+
+    for item in remaining_indices.iter() {
+        match item {
+            ExpandDims => continue,
+            TakeIndex { mut index } => {
+                if !have_array {
+                    index = resolve_index_unchecked(index, src.dim(axis as i32) as usize) as i32;
+                    starts[axis] = index;
+                    ends[axis] = index + 1;
+                    squeeze_needed = true;
+                }
+            }
+            Slice(range) => {
+                let size = src.dim(axis as i32);
+                starts[axis] = range.start(size);
+                ends[axis] = range.end(size);
+                strides[axis] = range.stride();
+            }
+            _ => unreachable!("Unexpected item in remaining_indices: {:?}", item),
+        }
+
+        axis += 1;
+    }
+
+    src = Cow::Owned(src.slice_device(&starts, &ends, &strides, stream));
+
+    // Unsqueeze handling
+    if remaining_indices.len() > ndim || squeeze_needed {
+        let mut new_shape = SmallVec::<[i32; 4]>::new();
+        let mut axis_ = 0;
+        for item in remaining_indices {
+            match item {
+                ExpandDims => new_shape.push(1),
+                TakeIndex { .. } => {
+                    if squeeze_needed {
+                        axis_ += 1;
+                    }
+                }
```

Comment:
This should be a bug right.. if squeeze is not needed, it should be going into the `_ =>` right?
Suggestion that would make that happen - tests pass in both cases, so I wonder.

```suggestion
                TakeIndex { .. } if squeeze_needed => axis_ += 1,
```

Reply:
What do you mean? Your code is doing the same thing

Reply:
Not exactly, in my version the if is in  the pattern matching itself, so if `squeeze_needed` is false, it will actually fallthrough into the `_ =>` case. Yours doesn't.

Reply:
Right, yours is actually the correct one

Reply:
Addressed in ec661bf335722cc03b7c0da398d6e683b0810bc1

Reply:
Do/Should we need to add a unit test here that would've caught this?

Reply:
> Do/Should we need to add a unit test here that would've caught this?

Maybe put it into the TODO list? Feel free to add one if you can come up with one.